Bump to cardano-api-8.20

locallycompact opened the pull request:
cardano-api: 8.11 -> 8.20.0
plutus: 1.7 -> 1.9

Drop BundledProtocolParameters and ProtocolParameters in favour of PParams LedgerEra

locallycompact committed:
Update iohk-nix and haskellNix

locallycompact committed:
Enable TypeOperators in hydra-node, hydra-cluster, hydra-cardano-api

locallycompact committed:
update CHaP and cabal index state

locallycompact committed:
cardano-api: 8.11 -> 8.20.0
plutus: 1.7 -> 1.9

Drop BundledProtocolParameters and ProtocolParameters in favour of PParams LedgerEra

ch1bo commented on hydra-cardano-api/src/Hydra/Cardano/Api.hs:
```diff
@@ -110,7 +109,7 @@ import Hydra.Cardano.Api.Prelude (
   Map,
   StandardCrypto,
-  ledgerEraVersion,
+  ledgerEraVersion, Proposal, VotingProcedures, LedgerProtocolParameters,
```
Please run `fourmolu` to format imports.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Fixture.hs:
```diff
@@ -45,22 +49,19 @@ testSeedInput = generateWith genTxIn 42
-    try (newLedgerEnv defaultPParams) >>= \case
-      Left (err :: ProtocolParametersConversionException) ->
-        error $ "Failed to create ledger env from fixture: " <> show err
-      Right env -> pure env
```
It's great to see this code removed!

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano/Configuration.hs:
```diff
@@ -90,22 +91,18 @@ newtype ProtocolParametersConversionException
-newLedgerEnv :: MonadThrow m => ProtocolParameters -> m (Ledger.LedgerEnv LedgerEra)
-newLedgerEnv protocolParams = do
-  case toLedgerPParams (shelleyBasedEra @Era) protocolParams of
-    Left err -> throwIO $ ProtocolParametersConversionException err
```
Should: remove now unused `ProtocolParametersConversionException`
Should: update the haddock of this function to not mention any exception throwing.

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano/Evaluate.hs:
```diff
@@ -86,6 +83,7 @@ import Test.QuickCheck (choose)
 import Test.QuickCheck.Gen (chooseWord64)
 import UntypedPlutusCore (UnrestrictedProgram (..))
 import qualified UntypedPlutusCore as UPLC
+import Hydra.Cardano.Api.Prelude (LedgerProtocolParameters(..))
```
Should: not import this from `Hydra.Cardano.Api.Prelude`, but from `Hydra.Cardano.Api`

ch1bo commented on hydra-node/hydra-node.cabal:
```diff
@@ -133,6 +134,7 @@ library
     , base16-bytestring
     , bech32
     , bytestring
+    , cardano-api:{internal}
```
We should not depend on `cardano-api` here, but instead only via `hydra-cardano-api`. Also, what are we using from the `internal` library?

ch1bo commented on hydra-node/src/Hydra/API/Server.hs:
```diff
@@ -64,7 +65,7 @@ withAPIServer ::
   Tracer IO APIServerLog ->
   Chain tx IO ->
-  ProtocolParameters ->
+  PParams LedgerEra ->
```
Seeing this, we _could_ make a pattern synonym for `PParams` part of the `hydra-cardano-api` for the same motivation: Having the latest era variant is what we usually want and it would make a single location able to change the era?

ch1bo commented on hydra-node/src/Hydra/Chain/CardanoClient.hs:
```diff
@@ -9,9 +9,11 @@ import Hydra.Prelude
 import qualified Data.Set as Set
 import Ouroboros.Consensus.HardFork.Combinator.AcrossEras (EraMismatch)
 import Test.QuickCheck (oneof)
+import Cardano.Api.ProtocolParameters (LedgerProtocolParameters(LedgerProtocolParameters))
```
Should: import this via `hydra-cardano-api` as this is our "anti corruption layer"

ch1bo commented on hydra-node/src/Hydra/Chain/Direct.hs:
```diff
@@ -154,9 +153,9 @@ mkTinyWallet tracer config = do
-    systemStart <- querySystemStart networkId nodeSocket QueryTip
+    walletUTxO <- Ledger.unUTxO . toLedgerUTxO <$> queryUTxO networkId nodeSocket (QueryAt point) [address]
+    pparams <- queryProtocolParameters networkId nodeSocket (QueryAt point)
+    systemStart <- querySystemStart networkId nodeSocket (QueryAt point)
```
Must: change this back to using `QueryTip` as we just fixed it.

This likely was due to a rebase conflict.

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano/Builder.hs:
```diff
@@ -6,6 +6,7 @@ import Hydra.Prelude
 
 import Data.Default (def)
 import qualified Data.Map as Map
+import Hydra.Cardano.Api.Prelude (LedgerProtocolParameters(..))
```
```suggestion
import Hydra.Cardano.Api (LedgerProtocolParameters(..))
```

ch1bo commented on hydra-test-utils/src/Test/Plutus/Validator.hs:
```diff
@@ -46,24 +47,20 @@ import Hydra.Cardano.Api (
   pattern TxInsCollateral,
   pattern TxOut,
  )
+import Hydra.Cardano.Api.Prelude (toAlonzoExUnits, LedgerProtocolParameters (LedgerProtocolParameters))
```
```suggestion
import Hydra.Cardano.Api (toLedgerExUnits, LedgerProtocolParameters (LedgerProtocolParameters))
```

ch1bo commented on nix/hydra/project.nix:
```diff
@@ -21,6 +21,7 @@ let
       # Reason: haskell.nix modules/overlays neds to be last
       # https://github.com/input-output-hk/haskell.nix/issues/1954
       haskellNix.overlay
+      iohk-nix.overlays.haskell-nix-crypto
```
This sounds like a duplicate to the above `iohk-nix.overlays.crypto`. Are both needed? If yes, a comment should explain why.

ch1bo replied:
@locallycompact Note that I changed this in https://github.com/input-output-hk/hydra/tree/5e38c1ab6c6c4e3f913cfc4d7ee3b308f525b842/

I think the static muslc build would fail if we re-order the overlays this way (see comment).

I also fixed the comments. It seems that `haskell-nix-crypto` is just containing pkg-config reconfiguration via haskell.nix. The libs are in the `crypto` overlay.

ch1bo reviewed (changes requested):
Thanks for wading through the dependency swamp!

IMO only two things MUST be done:

* [ ] Not undo the previous fix of querying `ChainTip` instead from point
* [ ] Update the `CHANGELOG.md` where at least the fact of new script hashes is warranted to be a `**BREAKING**` entry.

This PR would be even better if it would have a high-level description of the changes (see also our [Contribution Guidelines](https://github.com/input-output-hk/hydra/blob/master/CONTRIBUTING.md#creating-a-pull-request))

Also, there are many imports changing from `Hydra.Cardano.API (ProtocolParameters)` to `Cardano.Ledger.Core (PParams)` just to refer to the type (which result in additional dependencies on packages). I saw that the cardano-api does re-export `PParams` and we could make that easily accessible through `hydra-cardano-api`.

ch1bo commented:
The CI errors are actually interesting: https://github.com/input-output-hk/hydra/pull/1075/checks?check_run_id=16878199862

We do write a placeholder version into the binaries and it seems that the darwin build of some library resulted in another occurrence of the same placeholder! (I'm so glad I have put this check)

To fix you should update to something (still) unique [here](https://github.com/input-output-hk/hydra/blob/a92c73ce080fb22aea9c815b7343aa4ddbd160a5/nix/hydra/packages.nix#L25) and [here](https://github.com/input-output-hk/hydra/blob/a92c73ce080fb22aea9c815b7343aa4ddbd160a5/hydra-prelude/cbits/revision.c#L6-L9).

locallycompact committed:
Add overlays comment

locallycompact committed:
Update revision placeholders

ch1bo committed:
Remove redundant parenthesis and run fourmolu

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano/Evaluate.hs:
```diff
@@ -258,36 +247,28 @@ prepareTxScripts tx utxo = do
+      .~ Prices
+        { prSteps = fromJust $ boundRational $ 721 % 10000000
+        , prMem = fromJust $ boundRational $ 577 % 10000
+        }
```
@locallycompact Note that I removed some redundant parenthesis here

ch1bo committed:
Change overlay sequence and update comments

Maybe not needed anymore, but this honors the comment just above in
project.nix which was put in place to fix static muslc builds.

ch1bo reviewed (approved):
I pushed two comments to address some nitpicks I would have had and (I think) a fix to the static muslc builds (which we would have run into after merging..)

Approved now
